messaging/src/main/java/org/axonframework/messaging/deadletter/DeadLetterQueue.java

```diff
@@ -0,0 +1,229 @@
+ * of a {@link DeadLetter}. When retrieving letters through {@link #take(String)} for evaluation, they
+ * can be removed with {@link DeadLetter#acknowledge()} after successful evaluation. Upon failure, the letter may be
+ * reentered in the queue through {@link DeadLetter#requeue()}.
+ * <p>
+ * A callback can be configured through {@link #onAvailable(String, Runnable)} that is automatically invoked when
+ * dead-letters are released and thus ready to be taken. Letter sequences may be released earlier by invoking
+ * {@link #release(Predicate)}.
+ *
+ * @param <T> An implementation of {@link Message} that represent the dead-letter.
+ * @author Steven van Beelen
+ * @see QueueIdentifier
+ * @see DeadLetter
+ * @since 4.6.0
+ */
+public interface DeadLetterQueue<T extends Message<?>> {
+
+    /**
+     * Enqueues a {@link Message} to this queue. The {@code deadLetter} will be FIFO ordered with all other dead letters
+     * of the same {@code identifier}.
+     *
+     * @param identifier The identifier of the queue to store the {@code deadLetter} in.
+     * @param deadLetter The {@link Message} to enqueue.
+     * @param cause      The cause for enqueueing the given {@code deadLetter}.
+     * @return A {@link DeadLetter} representing the enqueued {@code deadLetter}.
+     * @throws DeadLetterQueueOverflowException Thrown when this queue is {@link #isFull(QueueIdentifier)} for the given
+     *                                          {@code identifier}.
+     */
+    DeadLetter<T> enqueue(@Nonnull QueueIdentifier identifier,
+                          @Nonnull T deadLetter,
+                          Throwable cause) throws DeadLetterQueueOverflowException;
+
+    /**
+     * Enqueue the given {@code message} only if there already are other {@link DeadLetter dead-letters} with the same
+     * {@code identifier} present in this queue.
+     *
+     * @param identifier The identifier of the queue to store the {@code deadLetter} in.
+     * @param message    The {@link Message} validated if it should be enqueued.
+     * @return An empty {@link Optional} if there are no {@link DeadLetter dead-letters} for the given
+     * {@code identifier}. A non-empty {@code Optional} is the result of the execution of
+     * {@link #enqueue(QueueIdentifier, Message, Throwable)}.
+     * @throws DeadLetterQueueOverflowException Thrown when this queue is {@link #isFull(QueueIdentifier)} for the given
+     *                                          {@code identifier}.
+     */
+    default Optional<DeadLetter<T>> enqueueIfPresent(@Nonnull QueueIdentifier identifier,
+                                                     @Nonnull T message) throws DeadLetterQueueOverflowException {
+        if (!contains(identifier)) {
+            return Optional.empty();
+        }
+
+        if (isFull(identifier)) {
+            throw new DeadLetterQueueOverflowException(
+                    "No room left to enqueue message [" + message + "] for identifier ["
```

Comment:
I suggest rewriting this exception message:
`Unable to enqueue message {messageType}. The maximum capacity of dead lettered messages has been reached.`

Other suggestions welcome :). The message should make clear that this is not an "Axon issue", but an issue with application capacity/behavior.